File: leetcode/1400/LC1461hasAllCodes.go
func hasAllCodes(s string, k int) bool {
	if len(s) < k {
		return false
	}
	left, right := 0, k
	set := make(map[string]struct{})
	set[s[left:right]] = struct{}{}

	for right < len(s) {
		right++
		left++
		set[s[left:right]] = struct{}{}
	}
	return len(set) == 1<<k
}

File: leetcode/1400/LC1461HasAllCodes.java
import java.util.HashSet;
import java.util.Set;

public class LC1461HasAllCodes {
    public boolean hasAllCodes(String s, int k) {
        if(s.length() < k){
            return false;
        }

        int left = 0, right = k;
        Set<String> set = new HashSet<>();
        set.add(s.substring(left, right));

        while(right < s.length()){
            right++;
            left++;
            set.add(s.substring(left, right));
        }

        return set.size() == Math.pow(2, k);
    }
}
